hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -231,39 +231,49 @@ protected void onPutIfAbsentAsyncInternal(K key, V value, Data keyData, Data val
     @Override
     protected void getAllInternal(Set<? extends K> keys, Collection<Data> dataKeys, ExpiryPolicy expiryPolicy,
                                   List<Object> resultingKeyValuePairs, long startNanos) {
-        Map<Object, Data> keyMap = null;
         if (serializeKeys) {
-            keyMap = toDataKeys(keys, dataKeys);
+            toDataKeysWithReservations(keys, dataKeys, null, null);
         }
-        Collection<?> ncKeys = serializeKeys ? dataKeys : new ArrayList<Object>(keys);
+        Collection<?> ncKeys = serializeKeys ? dataKeys : new LinkedList<K>(keys);
 
         populateResultFromNearCache(ncKeys, resultingKeyValuePairs);
         if (ncKeys.isEmpty()) {
             return;
         }
 
+        Map<Object, Long> reservations = createHashMap(ncKeys.size());
+        Map<Data, Object> reverseKeyMap = null;
         if (!serializeKeys) {
-            keyMap = toDataKeys(ncKeys, dataKeys);
+            reverseKeyMap = createHashMap(ncKeys.size());
+            toDataKeysWithReservations(ncKeys, dataKeys, reservations, reverseKeyMap);
+        } else {
+            createNearCacheReservations(ncKeys, reservations);
         }
 
-        Map<Object, Long> reservations = getNearCacheReservations(ncKeys, keyMap);
         try {
             int currentSize = resultingKeyValuePairs.size();
             super.getAllInternal(keys, dataKeys, expiryPolicy, resultingKeyValuePairs, startNanos);
```

Comment:
Out of curiosity - if we haven't fetched all keys from near cache, we load all of them from remote? Not just the missing ones?